sips/pending/_posts/2016-01-11-static-members.md

```diff
@@ -0,0 +1,151 @@
+
+## Use Cases
+
+Some JVM and JavaScript frameworks require classes to have specific static fields and/or methods.
+
+For example, classes extending `android.os.Parcelable` are required to have a static field named `CREATOR` of type `android.os.Parcelable$Creator`.
+
+Another example is using an [`AtomicReferenceFieldUpdater`](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/AtomicReferenceFieldUpdater.html).
+
+On the JavaScript side, one example is [Relay Route Definitions](https://facebook.github.io/relay/docs/guides-routes.html), whose subclasses must define static fields such as `queries`.
+Static methods and fields for JavaScript classes are one of the very few things (if not the only thing) that Scala.js "cannot do" at the moment, at least not declaratively.
+
+## Overview ##
+
+In order for method or field to be considered static it needs to be defined in an `object` and annotated `@static`.
+There is no special syntax proposed to access these members, they are accessed as if they were a member of defining objects with all appropriate access requirements for accessing them.
+
+For example:
+
+```scala
+{% highlight scala %}
+class Foo
+
+object Foo {
+  @static val x = 5
+  @static def bar(y: Int): Int = x + y
+}
+
+println(Foo.x)
+println(Foo.bar(12))
+{% endhighlight %}
+```
+
+Intuively, the presence of the `@static` annotation ensures that a field/method is declared as a static member of the companion class.
+For the JVM, the above would therefore look to other Java code as if it had been declared with the following Java code:
+
+```java
+{% highlight java %}
+class Foo {
+  public static int x = 5;
+  public static int bar(int y) {
+    return x + y;
+  }
+}
+{% endhighlight %}
+```
+
+In Scala.js, the `@static` annotation has no semantic effect in Scala objects, as they are not visible from JavaScript anyway (it could be used for optimizations).
+It has a semantic effect on Scala.js-defined JS classes, for example:
+
+```scala
+{% highlight scala %}
+@ScalaJSDefined
+class Foo extends js.Object
+
+@ScalaJSDefined
+object Foo extends js.Object {
+  @static val x = 5
+  @static def bar(y: Int): Int = x + y
+}
+{% endhighlight %}
+```
+
+would look to JavaScript code as if it had been declared with the following JavaScript code:
+
+```javascript
+{% highlight javascript %}
+class Foo extends Object {
+  static bar(y) {
+    return x + y;
+  }
+}
+Foo.x = 5; // in ES6, there is no declarative syntax for static fields yet
+{% endhighlight %}
+```
+
+## Comparisson with mirror classes ##
+
+Scalac currently generates static forwarders for fields and methods in top-level objects:
+
+```scala
+{% highlight scala %}
+object O {
+  val d = 1 
+  object I {
+    val f = 1
+  }
+}
+{% endhighlight %}
+```
+
+Under the proposed scheme users will be able to opt-in to have the field `f` defined in the inner object `I` emited as a static field. 
```

Comment:
The current implementation, if companion is missing, is going to generate static field in the `O.I$`.
This is used internally to implement lazy vals since https://github.com/lampepfl/dotty/pull/1226.
I'm not sure if I want to spec this, as this is subject to change.